alternator/ttl.cc

```diff
@@ -261,6 +263,187 @@ static future<> expire_item(service::storage_proxy& proxy,
+            std::optional<dht::token_range> ret = _intersecter->next();
+            if (ret) {
+                return ret;
+            }
+            // done with this range, go to next one
+            ++_range_it;
+            _intersecter = std::nullopt;
+        }
+        if (_range_it == _token_ranges.cend()) {
+            return std::nullopt;
+        }
+        _intersecter.emplace(_s->get_sharder(), *_range_it, this_shard_id());
+        // _intersecter.next() can't be empty, because we can assume that
+        // every vnode must intersect with every shard (we don't have vnodes
+        // that aren't serviced by some of the shards). If we didn't make this
+        // assumption, we would need a loop here, to potentially iterate
+        // through several rnages until finding one with an intersection.
+        std::optional<dht::token_range> ret = _intersecter->next();
+        assert(ret);
+        return ret;
+    }
+
+    // Same as next(), just return a partition_range instead of token_range
+    std::optional<dht::partition_range> next_partition_range() {
+        std::optional<dht::token_range> ret = next();
+        if (ret) {
+            return dht::to_partition_range(*ret);
+        } else {
+            return std::nullopt;
+        }
+    }
+};
+
+// Scan data in a list of token ranges in one table, looking for expired
+// items and deleting them.
+// Because of issue #9167, partition_ranges must have a single partition
+// for this code to work correctly.
+// FIXME: The long list of parameters below is ugly. This function should
+// calculate what it needs, or we can have an object with many of these
+// values pre-calculated in one function and then used in this one.
+static future<> scan_table_ranges(
+        service::storage_proxy& proxy,
+        schema_ptr s,
+        dht::partition_range_vector&& partition_ranges,
+        const bytes& column_name,
+        const std::optional<std::string>& member,
+        shared_ptr<cql3::selection::selection> selection,
+        service::query_state& query_state,
+        cql3::query_options& query_options,
+        lw_shared_ptr<query::read_command> command,
+        abort_source& abort_source)
+{
+    assert (partition_ranges.size() == 1); // otherwise issue #9167 will cause incorrect results.
+    auto p = service::pager::query_pagers::pager(s, selection, query_state, query_options, command, std::move(partition_ranges), nullptr);
+    while (!p->is_exhausted()) {
+        if (abort_source.abort_requested()) {
+            co_return;
+        }
+        // We don't to limit page size in number of rows because there is a
+        // builtin limit of the page's size in bytes. Setting this limit to 1
+        // is useful for debugging the paging code with moderate-size data.
+        uint32_t limit = std::numeric_limits<uint32_t>::max();
+        std::unique_ptr<cql3::result_set> rs = co_await p->fetch_page(limit, gc_clock::now(), executor::default_timeout());
```

Comment:
Is the default timeout appropriate?

We need an outer loop to restart in case we get an exception.

Reply:
Right. Adding FIXMEs for this.